Stable packages (work in progress)

steckerhalter opened the pull request:
don't merge yet, addresses #42

- [x] cache can deal with stable feature

steckerhalter commented:
@wasamasa here's an attempt at stable packages, can you test a bit?

wasamasa commented on quelpa.el:
```diff
@@ -388,7 +394,8 @@ If t, `quelpa' tries to do an upgrade.
       (pcase key
-        (:upgrade (setq quelpa-upgrade-p value))))
+        (:upgrade (setq quelpa-upgrade-p value))
+        (:stable (setq quelpa-stable-p value))))
```
Why is a "parsing" function setting a global value again? I wonder what I was thinking when I introduced that again.

steckerhalter replied:
it's good because it's shadowed, see below

wasamasa commented on quelpa.el:
```diff
@@ -472,6 +479,8 @@ the global var `quelpa-upgrade-p' is set to nil."
     ;; shadow `quelpa-upgrade-p' taking the default from the global var
     (let* ((quelpa-upgrade-p (if current-prefix-arg t quelpa-upgrade-p))
+           ;; shadow `quelpa-stable-p'
+           (quelpa-stable-p quelpa-stable-p)
```
How exactly does "shadowing" work here? Let-binding a variable to its value seems kind of silly.

steckerhalter replied:
the let binding makes it so that the globale variable is not overwritten by `setq` in the nested code, the default is taken from the global variable and as the plist is parsed it does not overwrite the global variable. I think that's kinda neat so you can do everything with one variable but in different scopes

wasamasa replied:
Oh right, let-binding them before allows the trick to work.  I still find it pretty convoluted compared to just using `or` on the global value and a plist retrieval though.

steckerhalter replied:
that's also an option, I think with the :upgrade property there would be a few special cases to consider but probably not for the :stable property.

steckerhalter commented:
weird build error now... I have to got now... will check that later

wasamasa commented:
[Uh-oh](https://github.com/milkypostman/melpa/pull/2640#issuecomment-89873379).

I'd suggest using a less troubled package instead.

steckerhalter commented:
I think it should work. I could install it locally using the tag. Personally I'm now already using the `next` branch and I tried installing the `stable` version which worked as well.

steckerhalter commented:
@wasamasa one thing that occured to me is that we need to check if the cache can deal with this stable feature

steckerhalter commented:
doesn't look too well with the cache:

```cl
((scss-mode :repo "antonj/scss-mode" :fetcher github)
 (anzu)
 (aes)
 (darktooth-theme)
 (noctilux-theme)
 (magit :fetcher github :repo "magit/magit" :commit "1.4.0")
 (magit :fetcher github :repo "magit/magit")
 (magit :fetcher github :repo "magit/magit" :files
        ("magit.el" "magit-bisect.el" "magit-blame.el" "magit-key-mode.el" "magit-popup.el" "magit-wip.el" "magit.texi" "with-editor.el" "AUTHORS.md" "README.md"))
 (ein)
 (magit :repo "magit/magit" :branch "next" :fetcher github)
 (outorg-export)
 (outorg)
 (sqlformat :fetcher github :repo "steckerhalter/sqlformat.el")
 (org-caldav)
 (org-caldav :repo "dengste/org-caldav" :fetcher github :files
             ("org-caldav.el"))
...
```
`stable` is of course not cached and it it also contains duplicate packages with different recipes
the cache should be fixed and make sure that there is only one `magit` for example, purgin older duplicates.

then we need to think how to add the stable information to the cache

wasamasa commented:
The issue with duplicate packages should be fixed now.  As for tracking the `:stable` keyword, I would probably push `:stable t` into the cache item at installation time. Installation of a package with unknown keyword-value pairs should work fine, they'd just need to be set via `quelpa-parse-plist` on the recipe.  I'd be rather concerned about the differences between specifying keyword-value pairs as plist following the recipe and in the recipe itself.

wasamasa commented:
The cache improvement is in `master` now, my remaining points still stand.

steckerhalter commented:
we could accept `:stable` both as part of the recipe and as option in our plist, if it is in our plist it would be pushed into the recipe so that it always ends up in the recipe in the end. so both of this will work:

```cl
(quelpa 'magit :stable t)
(quelpa '(magit :fetcher github :repo "magit/magit" :stable t))
```

wasamasa commented:
Once `:stable t` is in the recipe, it will end up in the cache. However, how shall that key-value combination be interpreted by the `quelpa` command? So far it is only honored if used as part of the PLIST argument because that one alters a locally bound variable, but what shall be done instead if it originates from the cache item?

steckerhalter commented:
I think the PLIST should only *add* it to the recipe so that in the end always the recipe would trigger the whole thing (the PLIST argument would overwrite the recipe's value if set). Something like that?

wasamasa commented:
Hmm, if you're doing it that way, that would mean `quelpa-parse-plist` shouldn't do anything when encountering `:stable`. It would also be differently handled compared to `:upgrade`.

I've added a change meanwhile that checks up if ARG contains `:stable` and if yes, sets the locally bound `quelpa-stable-p` to its value.

steckerhalter commented:
I think it's a priority question. how about: recipe > plist > `quelpa-stable-p`

so if the recipe specifies `:stable` it takes preference and so on?

but then we have the problem again that if we push `stable` into the recipe in the cache it would always have priority afterwards. hmm...

wasamasa commented:
Well, this would happen as `quelpa-stable-p` is interpreted first when let-binding it, then `quelpa-parse-plist` updates its bound value if something is present in PLIST and my change does it again when anything is present in the recipe (aka ARG).

I'd be fine with the way the cache is treated as it ensures that once someone chose building the stable version, upgrading would build the stable version as well until they change their mind.

steckerhalter commented:
yes you are right, it would actually be quite coherent. so we take that route then?

wasamasa commented:
I'd like to add at least a test for that first, but have a hard time thinking up how it would work out with ERT. Migrating to [buttercup](https://github.com/jorgenschaefer/emacs-buttercup) would be an option for testing the API thoroughly, but I'll leave that open for later, together with commits to make the logic of quelpa easier to understand (like, by abolishing all the shadowing going on currently).

Feel free to merge this.

steckerhalter commented:
it's still not pushing `stable` into the recipe (for the cache) nor considering `stable` if it is part of the recipe. ideally it would work like this:

if stable is part of the plist it would push it into the recipe
then when the recipe is processed it would either act on the stable value or - if that is not set - consider `quelpa-stable-p`

do you know what I mean?

wasamasa commented:
Try again please.

wasamasa commented:
I've devised the most unelegant test ever, that should make sure all three ways of building a stable package should work.

steckerhalter commented:
I've checked and this case:
```cl
(quelpa '(multiple-cursors :fetcher github :repo "magnars/multiple-cursors.el" :stable t))
```
the cache contains stable twice:

```cl
((multiple-cursors :fetcher github :repo "magnars/multiple-cursors.el" :stable t :stable t))
```

and this case is also a bit weird:

```cl
(let ((quelpa-stable-p t))
  (quelpa '(multiple-cursors :fetcher github :repo "magnars/multiple-cursors.el") :stable nil))

quelpa-cache
((multiple-cursors :fetcher github :repo "magnars/multiple-cursors.el")
 (multiple-cursors :fetcher github :repo "magnars/multiple-cursors.el"))
```

it shouldn't be twice I guess, not sure if the. I guess it doesn't matter if :stable nil is pushed into the recipe or not since it's only relevant if it is `t`

I added some documentation and regenerated the TOC

steckerhalter commented:
Moved Windows instructions out of the way since it is irrelevant for like 95% of `quelpa` users

wasamasa commented:
Any further findings, bugs, updates on this?

steckerhalter commented:
oh, i didn't see that you changed something again. will check again later.

steckerhalter commented:
Added some doc and a News section on top of the readme so that people get to know that we have added stable package support.

What I was wondering about is what happens if one sets the stable flag and specifies a commit or branch? Maybe there are other cases that could clash with stable packages?

wasamasa commented:
TIAS I guess.

steckerhalter commented:
anyways... I didn't find any more problems. since stable is additional it shouldn't affect the other stuff if it is not used. let's merge and see...
